py_trees/trees.py

```diff
@@ -192,21 +197,35 @@ def replace_subtree(self, unique_id, subtree):
                     return True
         return False
 
-    def setup(self, timeout):
+    def setup(self, timeout: float, visitor: visitors.VisitorBase=None):
         """
-         Relays to calling the :meth:`~py_trees.behaviour.Behaviour.setup` method
-         on the root behaviour. This in turn should get recursively called down through
-         the entire tree (if it does not, a behaviour with children has not
-         recursively called :meth:`~py_trees.behaviour.Behaviour.setup` on each of it's
-         children as it should).
+        Crawls across the tree calling :meth:`~py_trees.behaviour.Behaviour.setup`
+        on each behaviour.
+
+        Visitors can optionally be provided to provide a node-by-node analysis
+        on the result of each node's :meth:`~py_trees.behaviour.Behaviour.setup`
+        before the next node's :meth:`~py_trees.behaviour.Behaviour.setup` is called.
+        This is useful on trees with relatively long setup times to progressively
+        report out on the current status of the operation.
 
         Args:
-            timeout (:obj:`float`): time to wait (use common.Duration.INFINITE to block indefinitely)
+            timeout (:obj:`float`): time (s) to wait (use common.Duration.INFINITE to block indefinitely)
```

Comment:
BTW can also be None.

Reply:
Aha, totally forgot to handle the no timeout provided case. Fixed.